cli/src/cmd/cast/estimate.rs

```diff
@@ -0,0 +1,104 @@
+// cast estimate subcommands
+use crate::{
+    opts::{cast::parse_name_or_address, EthereumOpts},
+    utils::parse_ether_value,
+};
+use cast::{Cast, TxBuilder};
+use clap::Parser;
+use ethers::{
+    providers::{Middleware, Provider},
+    types::{NameOrAddress, U256},
+};
+use foundry_config::{Chain, Config};
+
+#[derive(Debug, Parser)]
+pub struct EstimateArgs {
+    #[clap(help = "The destination of the transaction.", parse(try_from_str = parse_name_or_address), value_name = "TO")]
+    to: Option<NameOrAddress>,
+    #[clap(help = "The signature of the function to call.", value_name = "SIG")]
+    sig: Option<String>,
+    #[clap(help = "The arguments of the function to call.", value_name = "ARGS")]
+    args: Vec<String>,
+    #[clap(
+        long,
+        help = "Ether to send in the transaction.",
+        long_help = r#"Ether to send in the transaction, either specified in wei, or as a string with a unit type.
+
+Examples: 1ether, 10gwei, 0.01ether"#,
+        parse(try_from_str = parse_ether_value),
+        value_name = "VALUE"
+    )]
+    value: Option<U256>,
+    #[clap(flatten)]
+    // TODO: We only need RPC URL and Etherscan API key here.
+    eth: EthereumOpts,
+    #[clap(subcommand)]
+    command: Option<EstimateSubcommands>,
+}
+
+#[derive(Debug, Parser)]
+pub enum EstimateSubcommands {
+    #[clap(name = "--create", about = "Estimate gas cost to deploy a smart contract")]
+    Create {
+        #[clap(help = "Bytecode of contract.", value_name = "CODE")]
+        code: String,
+        #[clap(help = "The signature of the function to call.", value_name = "SIG")]
```

Comment:
```suggestion
        #[clap(help = "The signature of the constructor to call.", value_name = "SIG")]
```

this should be the constructor, right?